prime/src/pages/BorrowActionsPage.tsx

```diff
@@ -415,6 +425,61 @@ export default function BorrowActionsPage() {
     [displayedMarginAccount, displayedUniswapPositions]
   );
 
+  // MARK: fetch uniswap fees earned
+  useEffect(() => {
+    let mounted = true;
+    if (marginAccount == null) {
+      return;
+    }
+    async function fetch(marginAccountLensContract: Contract, marginAccount: MarginAccount) {
+      const earnedFees: [string[], BigNumber[]] = await marginAccountLensContract.getUniswapFees(marginAccount.address);
+      const earnedFeesMap: UniswapPositionEarnedFees = {};
+      earnedFees[0].forEach((positionId, index) => {
+        earnedFeesMap[positionId] = {
+          token0FeesEarned: toBig(earnedFees[1][index * 2])
+            .div(10 ** marginAccount.token0.decimals)
+            .toNumber(),
+          token1FeesEarned: toBig(earnedFees[1][index * 2 + 1])
+            .div(10 ** marginAccount.token1.decimals)
+            .toNumber(),
+        };
+      });
+      if (mounted) {
+        setUniswapPositionEarnedFees(earnedFeesMap);
+      }
+    }
+    if (marginAccountLensContract && marginAccount) {
+      fetch(marginAccountLensContract, marginAccount);
+    }
+
+    return () => {
+      mounted = false;
+    };
+  }, [marginAccount, marginAccountLensContract]);
+
+  // MARK: compute uniswap fees earned
+  useEffect(() => {
+    if (!uniswapPositionEarnedFees || !displayedUniswapPositions || !marginAccount || !displayedMarginAccount) {
+      return;
+    }
+    const { token0, token1 } = marginAccount;
+    const earnedFeesValues = Object.values(uniswapPositionEarnedFees);
+    const token0FeesEarned = new Big(earnedFeesValues.reduce((p, c) => p + c.token0FeesEarned, 0));
+    const token1FeesEarned = new Big(earnedFeesValues.reduce((p, c) => p + c.token1FeesEarned, 0));
+    const token1OverToken0 = convertSqrtPriceX96(BigNumber.from(marginAccount.sqrtPriceX96.toFixed(0)));
+    const token0OverToken1 = token1OverToken0.pow(-1);
+    const tokenDecimalDifference = 10 ** (token1.decimals - token0.decimals);
+    const token1FeesEarnedInTermsOfToken0 = token0OverToken1.mul(token1FeesEarned).div(tokenDecimalDifference);
+    const token0FeesEarnedInTermsOfToken1 = token1OverToken0.mul(token0FeesEarned).div(tokenDecimalDifference);
+    const totalFeesEarned = isToken0Selected
+      ? token0FeesEarned.add(token1FeesEarnedInTermsOfToken0)
+      : token1FeesEarned.add(token0FeesEarnedInTermsOfToken1);
+    if (totalFeesEarned.gt(0)) {
+      const selectedTokenDecimals = isToken0Selected ? marginAccount.token0.decimals : marginAccount.token1.decimals;
+      setSwapFeesInputValue(totalFeesEarned.toFixed(selectedTokenDecimals));
+    }
+  }, [marginAccount, displayedMarginAccount, uniswapPositionEarnedFees, displayedUniswapPositions, isToken0Selected]);
```

Comment:
While this math is correct, it's not consistent with how we would compute similar values on-chain. On-chain we compute everything in terms of token1 to avoid having denominators > numerators and entering the scary realm of decimals.

I'd like to do the same here -- just compute total earned fees in terms of token1, and only convert to token0 (if necessary) at the very end. Instead of what you currently have, where there are essentially 2 parallel paths of the math.

Another thing to address in a future PR would be to break this out into its own custom hook. Maybe we can do this when we switch to GoodNumber.

Reply:
I just updated it, let me know if this is what you imagined.